### PR TITLE
Add native support to MacOS

### PR DESCRIPTION
**Pull request by pablogsal**

### CODE REVIEW

**Comment by yagodorea** [comment]
I cloned into the `native_macos` branch and tried to install, but got this error:

```log
    In file included from src/memray/_memray.cpp:844:
    In file included from src/memray/_memray/source.h:9:
    src/memray/_memray/lz4_stream.h:35:10: fatal error: 'lz4frame.h' file not found
    #include <lz4frame.h>
             ^~~~~~~~~~~~
    1 warning and 1 error generated.
    error: command '/usr/bin/gcc' failed with exit code 1
    [end of output]
```

Am I doing something wrong?

**Comment by pablogsal** [comment]
> Am I doing something wrong?

Yes, you need to ensure that the lz4 headers are available. For example, if you install lz4 with brew (`brew install lz4`) then you need to do `export CFLAGS="-I/opt/homebrew/opt/lz4/include"` before installing the library.

**Comment by yagodorea** [comment]
> > Am I doing something wrong?
> 
> Yes, you need to ensure that the lz4 headers are available. For example, if you install lz4 with brew (`brew install lz4`) then you need to do `export CFLAGS="-I$(brew --prefix lz4)/include"` ( before installing the library.

That did it, but some other errors like this arose (I'm using macOS 12.5):
```
    In file included from src/memray/_memray.cpp:863:
    src/memray/_memray/snapshot.h:107:61: error: 'value' is unavailable: introduced in macOS 10.13
                    removed_intervals.emplace_back(intersection.value(), value);
```

Maybe it's better that I wait for the release 😅 thanks for the help!

**Comment by pablogsal** [comment]
> > > Am I doing something wrong?
> > 
> > Yes, you need to ensure that the lz4 headers are available. For example, if you install lz4 with brew (`brew install lz4`) then you need to do `export CFLAGS="-I$(brew --prefix lz4)/include"` ( before installing the library.
> 
> That did it, but some other errors like this arose (I'm using macOS 12.5):
> ```
>     In file included from src/memray/_memray.cpp:863:
>     src/memray/_memray/snapshot.h:107:61: error: 'value' is unavailable: introduced in macOS 10.13
>                     removed_intervals.emplace_back(intersection.value(), value);
> ```
> 
> Maybe it's better that I wait for the release 😅 thanks for the help!

That's likely indicating you are using a old version of macOS. Even when we release we are going to require c++17 so you need macOS 11 or newer

**Comment by yagodorea** [comment]
> That's likely indicating you are using a old version of macOS

I thought of that too, but I think I just needed to set the env var `MACOSX_DEPLOYMENT_TARGET=12.5`

That said, I was able to successfully install it using these CFLAGS: `-I/opt/homebrew/include -L/opt/homebrew/lib`, but I can't run it now 🤡

```
  from ._memray import AllocationRecord
ImportError: dlopen(/...../memray/src/memray/_memray.cpython-38-darwin.so, 0x0002): symbol not found in flat namespace (_LZ4F_compressBegin)
```

**Comment by pablogsal** [comment]
> > That's likely indicating you are using a old version of macOS
> 
> I thought of that too, but I think I just needed to set the env var `MACOSX_DEPLOYMENT_TARGET=12.5`
> 
> That said, I was able to successfully install it using these CFLAGS: `-I/opt/homebrew/include -L/opt/homebrew/lib`, but I can't run it now 🤡
> 
> ```
>   from ._memray import AllocationRecord
> ImportError: dlopen(/...../memray/src/memray/_memray.cpython-38-darwin.so, 0x0002): symbol not found in flat namespace (_LZ4F_compressBegin)
> ```

That's because you are not linking against lz4. You need to probably also pass `-L` with the location of the lz4 shared object.

**Comment by pablogsal** [comment]
`-L/opt/homebrew/lib` looks wrong. It should be `-L$(brew --prefix lz4)/lib`

**Comment by yagodorea** [comment]
> @yagodorea You can check if the library is being linked by using `otool -L `. For example, in my MacBook machine I get:
> 
> ```
>  otool -L src/memray/_memray.cpython-310-darwin.so
> src/memray/_memray.cpython-310-darwin.so:
> 	/opt/homebrew/opt/lz4/lib/liblz4.1.dylib (compatibility version 1.0.0, current version 1.9.3)
> 	/usr/lib/libSystem.B.dylib (compatibility version 1.0.0, current version 1311.100.3)
> 	/usr/lib/libc++.1.dylib (compatibility version 1.0.0, current version 1300.23.0)
> ```
> 
> Notice that I have `/opt/homebrew/opt/lz4/lib/liblz4.1.dylib` which means is using that shared object for the lz4 symbols.

Yeah even with `CFLAGS="-I$(brew --prefix lz4)/include -L$(brew --prefix lz4)/lib"` it's not linking the lz4 lib:
```
> otool -L src/memray/_memray.cpython-38-darwin.so
src/memray/_memray.cpython-38-darwin.so:
	/usr/lib/libSystem.B.dylib (compatibility version 1.0.0, current version 1311.100.3)
	@rpath/libc++.1.dylib (compatibility version 1.0.0, current version 1.0.0)
```

**Comment by pablogsal** [comment]
> Yeah even with `CFLAGS="-I$(brew --prefix lz4)/include -L$(brew --prefix lz4)/lib"` it's not linking the lz4 lib:

Can you print the compiler commands? (The output of `make` or `pip install -e . -vvvv`)

**Comment by pablogsal** [comment]
Something very weird is happening because we link against lz4 always:

[```

```](https://github.com/bloomberg/memray/blob/3291202571471b1075b2a0a1ffa852ba1b52c0cc/setup.py#L225)

**Comment by yagodorea** [comment]
> > Yeah even with `CFLAGS="-I$(brew --prefix lz4)/include -L$(brew --prefix lz4)/lib"` it's not linking the lz4 lib:
> 
> Can you print the compiler commands? (The output of `make` or `pip install -e . -vvvv`)

Got some lz4 linking warnings with -vvvv, and adding `DYLD_LIBRARY_PATH="$(brew --prefix lz4)/lib"` didn't work. 

`ld: warning: ignoring file /opt/homebrew/opt/lz4/lib/liblz4.dylib, building for macOS-x86_64 but attempting to link with file built for macOS-arm64`

```
gcc -Wno-unused-result -Wsign-compare -Wunreachable-code -DNDEBUG -g -fwrapv -O3 -Wall -Wstrict-prototypes -I/opt/anaconda3/envs/alvpy/include -arch x86_64 -I/opt/anaconda3/envs/alvpy/include -arch x86_64 -I/opt/homebrew/opt/lz4/include -L/opt/homebrew/opt/lz4/lib -D_FILE_OFFSET_BITS=64 -DUSE_MEMRAY_TLS_MODEL=1 -Isrc/memray/_memray -Isrc -I/Users/yagodorea/code/memray/src/vendor/libbacktrace/install/include -I/opt/anaconda3/envs/alvpy/include/python3.8 -c src/memray/_memray/compat.cpp -o build/temp.macosx-10.9-x86_64-cpython-38/src/memray/_memray/compat.o -std=c++17 -Wall -I/opt/homebrew/opt/lz4/include -L/opt/homebrew/opt/lz4/lib -flto
    clang: warning: argument unused during compilation: '-L/opt/homebrew/opt/lz4/lib' [-Wunused-command-line-argument]
    clang: warning: argument unused during compilation: '-L/opt/homebrew/opt/lz4/lib' [-Wunused-command-line-argument]
    gcc -Wno-unused-result -Wsign-compare -Wunreachable-code -DNDEBUG -g -fwrapv -O3 -Wall -Wstrict-prototypes -I/opt/anaconda3/envs/alvpy/include -arch x86_64 -I/opt/anaconda3/envs/alvpy/include -arch x86_64 -I/opt/homebrew/opt/lz4/include -L/opt/homebrew/opt/lz4/lib -D_FILE_OFFSET_BITS=64 -DUSE_MEMRAY_TLS_MODEL=1 -Isrc/memray/_memray -Isrc -I/Users/yagodorea/code/memray/src/vendor/libbacktrace/install/include -I/opt/anaconda3/envs/alvpy/include/python3.8 -c src/memray/_memray/hooks.cpp -o build/temp.macosx-10.9-x86_64-cpython-38/src/memray/_memray/hooks.o -std=c++17 -Wall -I/opt/homebrew/opt/lz4/include -L/opt/homebrew/opt/lz4/lib -flto
    clang: warning: argument unused during compilation: '-L/opt/homebrew/opt/lz4/lib' [-Wunused-command-line-argument]
    clang: warning: argument unused during compilation: '-L/opt/homebrew/opt/lz4/lib' [-Wunused-command-line-argument]
    gcc -Wno-unused-result -Wsign-compare -Wunreachable-code -DNDEBUG -g -fwrapv -O3 -Wall -Wstrict-prototypes -I/opt/anaconda3/envs/alvpy/include -arch x86_64 -I/opt/anaconda3/envs/alvpy/include -arch x86_64 -I/opt/homebrew/opt/lz4/include -L/opt/homebrew/opt/lz4/lib -D_FILE_OFFSET_BITS=64 -DUSE_MEMRAY_TLS_MODEL=1 -Isrc/memray/_memray -Isrc -I/Users/yagodorea/code/memray/src/vendor/libbacktrace/install/include -I/opt/anaconda3/envs/alvpy/include/python3.8 -c src/memray/_memray/logging.cpp -o build/temp.macosx-10.9-x86_64-cpython-38/src/memray/_memray/logging.o -std=c++17 -Wall -I/opt/homebrew/opt/lz4/include -L/opt/homebrew/opt/lz4/lib -flto
    clang: warning: argument unused during compilation: '-L/opt/homebrew/opt/lz4/lib' [-Wunused-command-line-argument]
    clang: warning: argument unused during compilation: '-L/opt/homebrew/opt/lz4/lib' [-Wunused-command-line-argument]
    gcc -Wno-unused-result -Wsign-compare -Wunreachable-code -DNDEBUG -g -fwrapv -O3 -Wall -Wstrict-prototypes -I/opt/anaconda3/envs/alvpy/include -arch x86_64 -I/opt/anaconda3/envs/alvpy/include -arch x86_64 -I/opt/homebrew/opt/lz4/include -L/opt/homebrew/opt/lz4/lib -D_FILE_OFFSET_BITS=64 -DUSE_MEMRAY_TLS_MODEL=1 -Isrc/memray/_memray -Isrc -I/Users/yagodorea/code/memray/src/vendor/libbacktrace/install/include -I/opt/anaconda3/envs/alvpy/include/python3.8 -c src/memray/_memray/macho_shenanigans.cpp -o build/temp.macosx-10.9-x86_64-cpython-38/src/memray/_memray/macho_shenanigans.o -std=c++17 -Wall -I/opt/homebrew/opt/lz4/include -L/opt/homebrew/opt/lz4/lib -flto
    clang: warning: argument unused during compilation: '-L/opt/homebrew/opt/lz4/lib' [-Wunused-command-line-argument]
    clang: warning: argument unused during compilation: '-L/opt/homebrew/opt/lz4/lib' [-Wunused-command-line-argument]
    gcc -Wno-unused-result -Wsign-compare -Wunreachable-code -DNDEBUG -g -fwrapv -O3 -Wall -Wstrict-prototypes -I/opt/anaconda3/envs/alvpy/include -arch x86_64 -I/opt/anaconda3/envs/alvpy/include -arch x86_64 -I/opt/homebrew/opt/lz4/include -L/opt/homebrew/opt/lz4/lib -D_FILE_OFFSET_BITS=64 -DUSE_MEMRAY_TLS_MODEL=1 -Isrc/memray/_memray -Isrc -I/Users/yagodorea/code/memray/src/vendor/libbacktrace/install/include -I/opt/anaconda3/envs/alvpy/include/python3.8 -c src/memray/_memray/native_resolver.cpp -o build/temp.macosx-10.9-x86_64-cpython-38/src/memray/_memray/native_resolver.o -std=c++17 -Wall -I/opt/homebrew/opt/lz4/include -L/opt/homebrew/opt/lz4/lib -flto
    clang: warning: argument unused during compilation: '-L/opt/homebrew/opt/lz4/lib' [-Wunused-command-line-argument]
    clang: warning: argument unused during compilation: '-L/opt/homebrew/opt/lz4/lib' [-Wunused-command-line-argument]
    gcc -Wno-unused-result -Wsign-compare -Wunreachable-code -DNDEBUG -g -fwrapv -O3 -Wall -Wstrict-prototypes -I/opt/anaconda3/envs/alvpy/include -arch x86_64 -I/opt/anaconda3/envs/alvpy/include -arch x86_64 -I/opt/homebrew/opt/lz4/include -L/opt/homebrew/opt/lz4/lib -D_FILE_OFFSET_BITS=64 -DUSE_MEMRAY_TLS_MODEL=1 -Isrc/memray/_memray -Isrc -I/Users/yagodorea/code/memray/src/vendor/libbacktrace/install/include -I/opt/anaconda3/envs/alvpy/include/python3.8 -c src/memray/_memray/python_helpers.cpp -o build/temp.macosx-10.9-x86_64-cpython-38/src/memray/_memray/python_helpers.o -std=c++17 -Wall -I/opt/homebrew/opt/lz4/include -L/opt/homebrew/opt/lz4/lib -flto
    clang: warning: argument unused during compilation: '-L/opt/homebrew/opt/lz4/lib' [-Wunused-command-line-argument]
    clang: warning: argument unused during compilation: '-L/opt/homebrew/opt/lz4/lib' [-Wunused-command-line-argument]
    gcc -Wno-unused-result -Wsign-compare -Wunreachable-code -DNDEBUG -g -fwrapv -O3 -Wall -Wstrict-prototypes -I/opt/anaconda3/envs/alvpy/include -arch x86_64 -I/opt/anaconda3/envs/alvpy/include -arch x86_64 -I/opt/homebrew/opt/lz4/include -L/opt/homebrew/opt/lz4/lib -D_FILE_OFFSET_BITS=64 -DUSE_MEMRAY_TLS_MODEL=1 -Isrc/memray/_memray -Isrc -I/Users/yagodorea/code/memray/src/vendor/libbacktrace/install/include -I/opt/anaconda3/envs/alvpy/include/python3.8 -c src/memray/_memray/record_reader.cpp -o build/temp.macosx-10.9-x86_64-cpython-38/src/memray/_memray/record_reader.o -std=c++17 -Wall -I/opt/homebrew/opt/lz4/include -L/opt/homebrew/opt/lz4/lib -flto
    clang: warning: argument unused during compilation: '-L/opt/homebrew/opt/lz4/lib' [-Wunused-command-line-argument]
    clang: warning: argument unused during compilation: '-L/opt/homebrew/opt/lz4/lib' [-Wunused-command-line-argument]
    gcc -Wno-unused-result -Wsign-compare -Wunreachable-code -DNDEBUG -g -fwrapv -O3 -Wall -Wstrict-prototypes -I/opt/anaconda3/envs/alvpy/include -arch x86_64 -I/opt/anaconda3/envs/alvpy/include -arch x86_64 -I/opt/homebrew/opt/lz4/include -L/opt/homebrew/opt/lz4/lib -D_FILE_OFFSET_BITS=64 -DUSE_MEMRAY_TLS_MODEL=1 -Isrc/memray/_memray -Isrc -I/Users/yagodorea/code/memray/src/vendor/libbacktrace/install/include -I/opt/anaconda3/envs/alvpy/include/python3.8 -c src/memray/_memray/record_writer.cpp -o build/temp.macosx-10.9-x86_64-cpython-38/src/memray/_memray/record_writer.o -std=c++17 -Wall -I/opt/homebrew/opt/lz4/include -L/opt/homebrew/opt/lz4/lib -flto
    clang: warning: argument unused during compilation: '-L/opt/homebrew/opt/lz4/lib' [-Wunused-command-line-argument]
    clang: warning: argument unused during compilation: '-L/opt/homebrew/opt/lz4/lib' [-Wunused-command-line-argument]
    gcc -Wno-unused-result -Wsign-compare -Wunreachable-code -DNDEBUG -g -fwrapv -O3 -Wall -Wstrict-prototypes -I/opt/anaconda3/envs/alvpy/include -arch x86_64 -I/opt/anaconda3/envs/alvpy/include -arch x86_64 -I/opt/homebrew/opt/lz4/include -L/opt/homebrew/opt/lz4/lib -D_FILE_OFFSET_BITS=64 -DUSE_MEMRAY_TLS_MODEL=1 -Isrc/memray/_memray -Isrc -I/Users/yagodorea/code/memray/src/vendor/libbacktrace/install/include -I/opt/anaconda3/envs/alvpy/include/python3.8 -c src/memray/_memray/records.cpp -o build/temp.macosx-10.9-x86_64-cpython-38/src/memray/_memray/records.o -std=c++17 -Wall -I/opt/homebrew/opt/lz4/include -L/opt/homebrew/opt/lz4/lib -flto
    clang: warning: argument unused during compilation: '-L/opt/homebrew/opt/lz4/lib' [-Wunused-command-line-argument]
    clang: warning: argument unused during compilation: '-L/opt/homebrew/opt/lz4/lib' [-Wunused-command-line-argument]
    gcc -Wno-unused-result -Wsign-compare -Wunreachable-code -DNDEBUG -g -fwrapv -O3 -Wall -Wstrict-prototypes -I/opt/anaconda3/envs/alvpy/include -arch x86_64 -I/opt/anaconda3/envs/alvpy/include -arch x86_64 -I/opt/homebrew/opt/lz4/include -L/opt/homebrew/opt/lz4/lib -D_FILE_OFFSET_BITS=64 -DUSE_MEMRAY_TLS_MODEL=1 -Isrc/memray/_memray -Isrc -I/Users/yagodorea/code/memray/src/vendor/libbacktrace/install/include -I/opt/anaconda3/envs/alvpy/include/python3.8 -c src/memray/_memray/sink.cpp -o build/temp.macosx-10.9-x86_64-cpython-38/src/memray/_memray/sink.o -std=c++17 -Wall -I/opt/homebrew/opt/lz4/include -L/opt/homebrew/opt/lz4/lib -flto
    clang: warning: argument unused during compilation: '-L/opt/homebrew/opt/lz4/lib' [-Wunused-command-line-argument]
    clang: warning: argument unused during compilation: '-L/opt/homebrew/opt/lz4/lib' [-Wunused-command-line-argument]
    gcc -Wno-unused-result -Wsign-compare -Wunreachable-code -DNDEBUG -g -fwrapv -O3 -Wall -Wstrict-prototypes -I/opt/anaconda3/envs/alvpy/include -arch x86_64 -I/opt/anaconda3/envs/alvpy/include -arch x86_64 -I/opt/homebrew/opt/lz4/include -L/opt/homebrew/opt/lz4/lib -D_FILE_OFFSET_BITS=64 -DUSE_MEMRAY_TLS_MODEL=1 -Isrc/memray/_memray -Isrc -I/Users/yagodorea/code/memray/src/vendor/libbacktrace/install/include -I/opt/anaconda3/envs/alvpy/include/python3.8 -c src/memray/_memray/snapshot.cpp -o build/temp.macosx-10.9-x86_64-cpython-38/src/memray/_memray/snapshot.o -std=c++17 -Wall -I/opt/homebrew/opt/lz4/include -L/opt/homebrew/opt/lz4/lib -flto
    clang: warning: argument unused during compilation: '-L/opt/homebrew/opt/lz4/lib' [-Wunused-command-line-argument]
    clang: warning: argument unused during compilation: '-L/opt/homebrew/opt/lz4/lib' [-Wunused-command-line-argument]
    gcc -Wno-unused-result -Wsign-compare -Wunreachable-code -DNDEBUG -g -fwrapv -O3 -Wall -Wstrict-prototypes -I/opt/anaconda3/envs/alvpy/include -arch x86_64 -I/opt/anaconda3/envs/alvpy/include -arch x86_64 -I/opt/homebrew/opt/lz4/include -L/opt/homebrew/opt/lz4/lib -D_FILE_OFFSET_BITS=64 -DUSE_MEMRAY_TLS_MODEL=1 -Isrc/memray/_memray -Isrc -I/Users/yagodorea/code/memray/src/vendor/libbacktrace/install/include -I/opt/anaconda3/envs/alvpy/include/python3.8 -c src/memray/_memray/socket_reader_thread.cpp -o build/temp.macosx-10.9-x86_64-cpython-38/src/memray/_memray/socket_reader_thread.o -std=c++17 -Wall -I/opt/homebrew/opt/lz4/include -L/opt/homebrew/opt/lz4/lib -flto
    clang: warning: argument unused during compilation: '-L/opt/homebrew/opt/lz4/lib' [-Wunused-command-line-argument]
    clang: warning: argument unused during compilation: '-L/opt/homebrew/opt/lz4/lib' [-Wunused-command-line-argument]
    gcc -Wno-unused-result -Wsign-compare -Wunreachable-code -DNDEBUG -g -fwrapv -O3 -Wall -Wstrict-prototypes -I/opt/anaconda3/envs/alvpy/include -arch x86_64 -I/opt/anaconda3/envs/alvpy/include -arch x86_64 -I/opt/homebrew/opt/lz4/include -L/opt/homebrew/opt/lz4/lib -D_FILE_OFFSET_BITS=64 -DUSE_MEMRAY_TLS_MODEL=1 -Isrc/memray/_memray -Isrc -I/Users/yagodorea/code/memray/src/vendor/libbacktrace/install/include -I/opt/anaconda3/envs/alvpy/include/python3.8 -c src/memray/_memray/source.cpp -o build/temp.macosx-10.9-x86_64-cpython-38/src/memray/_memray/source.o -std=c++17 -Wall -I/opt/homebrew/opt/lz4/include -L/opt/homebrew/opt/lz4/lib -flto
    clang: warning: argument unused during compilation: '-L/opt/homebrew/opt/lz4/lib' [-Wunused-command-line-argument]
    clang: warning: argument unused during compilation: '-L/opt/homebrew/opt/lz4/lib' [-Wunused-command-line-argument]
    gcc -Wno-unused-result -Wsign-compare -Wunreachable-code -DNDEBUG -g -fwrapv -O3 -Wall -Wstrict-prototypes -I/opt/anaconda3/envs/alvpy/include -arch x86_64 -I/opt/anaconda3/envs/alvpy/include -arch x86_64 -I/opt/homebrew/opt/lz4/include -L/opt/homebrew/opt/lz4/lib -D_FILE_OFFSET_BITS=64 -DUSE_MEMRAY_TLS_MODEL=1 -Isrc/memray/_memray -Isrc -I/Users/yagodorea/code/memray/src/vendor/libbacktrace/install/include -I/opt/anaconda3/envs/alvpy/include/python3.8 -c src/memray/_memray/tracking_api.cpp -o build/temp.macosx-10.9-x86_64-cpython-38/src/memray/_memray/tracking_api.o -std=c++17 -Wall -I/opt/homebrew/opt/lz4/include -L/opt/homebrew/opt/lz4/lib -flto
    clang: warning: argument unused during compilation: '-L/opt/homebrew/opt/lz4/lib' [-Wunused-command-line-argument]
    clang: warning: argument unused during compilation: '-L/opt/homebrew/opt/lz4/lib' [-Wunused-command-line-argument]
    g++ -bundle -undefined dynamic_lookup -L/opt/anaconda3/envs/alvpy/lib -arch x86_64 -L/opt/anaconda3/envs/alvpy/lib -arch x86_64 -I/opt/homebrew/opt/lz4/include -L/opt/homebrew/opt/lz4/lib build/temp.macosx-10.9-x86_64-cpython-38/src/memray/_memray.o build/temp.macosx-10.9-x86_64-cpython-38/src/memray/_memray/compat.o build/temp.macosx-10.9-x86_64-cpython-38/src/memray/_memray/hooks.o build/temp.macosx-10.9-x86_64-cpython-38/src/memray/_memray/logging.o build/temp.macosx-10.9-x86_64-cpython-38/src/memray/_memray/macho_shenanigans.o build/temp.macosx-10.9-x86_64-cpython-38/src/memray/_memray/native_resolver.o build/temp.macosx-10.9-x86_64-cpython-38/src/memray/_memray/python_helpers.o build/temp.macosx-10.9-x86_64-cpython-38/src/memray/_memray/record_reader.o build/temp.macosx-10.9-x86_64-cpython-38/src/memray/_memray/record_writer.o build/temp.macosx-10.9-x86_64-cpython-38/src/memray/_memray/records.o build/temp.macosx-10.9-x86_64-cpython-38/src/memray/_memray/sink.o build/temp.macosx-10.9-x86_64-cpython-38/src/memray/_memray/snapshot.o build/temp.macosx-10.9-x86_64-cpython-38/src/memray/_memray/socket_reader_thread.o build/temp.macosx-10.9-x86_64-cpython-38/src/memray/_memray/source.o build/temp.macosx-10.9-x86_64-cpython-38/src/memray/_memray/tracking_api.o -L/Users/yagodorea/code/memray/src/vendor/libbacktrace/install/lib -llz4 -ldl -o build/lib.macosx-10.9-x86_64-cpython-38/memray/_memray.cpython-38-darwin.so -std=c++17 -lbacktrace -flto
    ld: warning: ignoring file /opt/homebrew/opt/lz4/lib/liblz4.dylib, building for macOS-x86_64 but attempting to link with file built for macOS-arm64
    building 'memray._test_utils' extension
    gcc -Wno-unused-result -Wsign-compare -Wunreachable-code -DNDEBUG -g -fwrapv -O3 -Wall -Wstrict-prototypes -I/opt/anaconda3/envs/alvpy/include -arch x86_64 -I/opt/anaconda3/envs/alvpy/include -arch x86_64 -I/opt/homebrew/opt/lz4/include -L/opt/homebrew/opt/lz4/lib -D_FILE_OFFSET_BITS=64 -DUSE_MEMRAY_TLS_MODEL=1 -Isrc/memray/_memray -I/opt/anaconda3/envs/alvpy/include/python3.8 -c src/memray/_memray_test_utils.cpp -o build/temp.macosx-10.9-x86_64-cpython-38/src/memray/_memray_test_utils.o -std=c++17 -Wall -I/opt/homebrew/opt/lz4/include -L/opt/homebrew/opt/lz4/lib -flto
    clang: warning: argument unused during compilation: '-L/opt/homebrew/opt/lz4/lib' [-Wunused-command-line-argument]
    clang: warning: argument unused during compilation: '-L/opt/homebrew/opt/lz4/lib' [-Wunused-command-line-argument]
```

**Comment by pablogsal** [comment]
Ah definitely this is the problem:

>     ld: warning: ignoring file /opt/homebrew/opt/lz4/lib/liblz4.dylib, building for macOS-x86_64 but attempting to link with file built for macOS-arm64

There is something wrong with your compiler chain because is building for an architecture and compiling for another. I assume you are on an M1 laptop, no?

**Comment by yagodorea** [comment]
> Ah definitely this is the problem:
> 
> > ```
> > ld: warning: ignoring file /opt/homebrew/opt/lz4/lib/liblz4.dylib, building for macOS-x86_64 but attempting to link with file built for macOS-arm64
> > ```
> 
> There is something wrong with your compiler chain because is building for an architecture and compiling for another. I assume you are on an M1 laptop, no?

Yes, an M1 Macbook Pro, but I don't usually mess with C so I don't think I could have misconfigured anything haha, unless some third party tool did it.

```
> g++ --version
Apple clang version 13.1.6 (clang-1316.0.21.2.5)
Target: arm64-apple-darwin21.6.0
Thread model: posix
InstalledDir: /Library/Developer/CommandLineTools/usr/bin
> lz4 --version
*** LZ4 command line interface 64-bits v1.9.3, by Yann Collet ***
```

Is my compiler trying to build memray for x86?

**Comment by pablogsal** [comment]
Somehow you are compiling for INTEL (`-arch x86_64`). Maybe is because your Python doesn't have for apple silicon. What's the output of `python -m sysconfig | grep arch` ?

I see you are using `anaconda3`. I don't know if anaconda is compiling Python with M1 support because we added that upstream in 3.9 for the first time.

In that case, you are out of luck because you cannot link a amd64 library (`/opt/homebrew/opt/lz4/lib/liblz4.dylib`) into an intel program (the one you are just compiling because your Python seems to be intel compiled). You should install a Python version compiled for Apple silicon like the one we offer upstream:

https://www.python.org/ftp/python/3.10.6/python-3.10.6-macos11.pkg

**Comment by yagodorea** [comment]
> I see you are using `anaconda3`. I don't know if anaconda is compiling Python with M1 support because we added that upstream in 3.9 for the first time.

That's it, all my conda env libs are compiled for x86 🤦‍♂️

Thanks for helping me through this rabbit hole, haha! I'll consider refactoring my team's env and then get back to this.

File: src/vendor/libbacktrace_2446c66076480ce07a6bd868badcbceb3eeecc2e_patch.diff [hunk 1]
```diff
@@ -118,9 +115,120 @@ index bb481f3..2b056be 100644
...
+ #define MACH_O_MH_DYLIB		0x06
++#define MACH_O_MH_BUNDLE	0x08
+ #define MACH_O_MH_DSYM		0x0a
+ 
+ /* A component of a fat file.  A fat file starts with a
+@@ -278,6 +279,11 @@ struct macho_nlist_64
+ #define MACH_O_N_TYPE	0x0e	/* Mask for type bits */
+ #define MACH_O_N_STAB	0xe0	/* Stabs debugging symbol */
+ 
++#define MACH_O_N_FUN	0xe0	/* Stabs debugging symbol */
```
Comment (godlygeek) [inline src/vendor/libbacktrace_2446c66076480ce07a6bd868badcbceb3eeecc2e_patch.diff:152]: Hm, I looked at this code above - seems like you've checked in patched code _and_ the patch for that code?

File: src/vendor/libbacktrace_2446c66076480ce07a6bd868badcbceb3eeecc2e_patch.diff [hunk 2]
```diff
@@ -118,9 +115,120 @@ index bb481f3..2b056be 100644
...
++#define MACH_O_GLOBAL   0x0f    /* global symbol */
++#define MACH_O_N_FNAME  0x26    /* static symbol: name,,n_sect,type,address */
++#define MACH_O_N_FUN    0x24    /* procedure: name,,n_sect,linenumber,address */
```
Comment (godlygeek) [inline src/vendor/libbacktrace_2446c66076480ce07a6bd868badcbceb3eeecc2e_patch.diff:155]: Wait, this can't possibly work - you defined this macro on 152, and you're re-defining it again on 155, without an intervening `#undef`

File: src/vendor/libbacktrace/macho.c [hunk 3]
```diff
@@ -278,6 +279,11 @@ struct macho_nlist_64
 #define MACH_O_N_TYPE	0x0e	/* Mask for type bits */
 #define MACH_O_N_STAB	0xe0	/* Stabs debugging symbol */
 
+#define MACH_O_N_FUN	0xe0	/* Stabs debugging symbol */
```
[the remark this answers is not in the document]
Comment (pablogsal) [reply src/vendor/libbacktrace/macho.c:282]: No, the second one is the correct one. Good catch

File: src/vendor/libbacktrace_2446c66076480ce07a6bd868badcbceb3eeecc2e_patch.diff (hunk 2 above)
Comment (pablogsal) [reply src/vendor/libbacktrace_2446c66076480ce07a6bd868badcbceb3eeecc2e_patch.diff:155]: Yeah, the first macro should be removed

File: src/vendor/libbacktrace_2446c66076480ce07a6bd868badcbceb3eeecc2e_patch.diff (hunk 1 above)
Comment (pablogsal) [reply src/vendor/libbacktrace_2446c66076480ce07a6bd868badcbceb3eeecc2e_patch.diff:152]: Yes, that's what we want: to update our vendored lib and the patch we used

Comment (godlygeek) [reply src/vendor/libbacktrace_2446c66076480ce07a6bd868badcbceb3eeecc2e_patch.diff]: Ah, my bad - I was thinking that we applied patches at build time, not that we checked in patched files along with the patch that generated them for use when we re-generate the vendored dependency.

File: tests/integration/native_extension/native_ext.c [hunk 4]
```diff
@@ -24,7 +26,7 @@ __attribute__((noinline)) static void foo() {
 }
 
 PyObject*
-run_simple(PyObject*, PyObject*)
+run_simple(PyObject* mod , PyObject* arg)
```
Comment (godlygeek) [inline tests/integration/native_extension/native_ext.c:29]: Why is this change needed? Is it actually somehow relevant to getting the test suite to pass on Mac?

Oh - does the C compiler that's being used on Mac complain if a parameter has no name?

File: tests/integration/test_main.py [hunk 5]
```diff
@@ -456,23 +456,19 @@ def test_successful_parse(self, tmp_path):
         # GIVEN
         record_types = [
             "ALLOCATION",
+            "ALLOCATION_WITH_NATIVE",
```
Comment (godlygeek) [inline tests/integration/test_main.py:459]: `ALLOCATION_WITH_NATIVE` is listed twice (both first and last line of the diff hunk)

File: tests/integration/test_tracking.py [hunk 6]
```diff
@@ -739,6 +737,7 @@ def test_partial_munmap_gap(self, tmp_path):
         peak_memory = sum(x.size for x in peak_allocations)
         assert peak_memory == 12 * PAGE_SIZE
 
+    @skip_if_macos
```
Comment (godlygeek) [inline tests/integration/test_tracking.py:740]: Hm, what's the reason for this test to need to be skipped? Are we not able to reliably get adjacent pages out of mmap on Mac?

**Review by godlygeek — CHANGES_REQUESTED** [review]

File: tests/integration/native_extension/native_ext.c (hunk 4 above)
Comment (pablogsal) [reply tests/integration/native_extension/native_ext.c:29]: The c compiler in the Linux images is old enough that complains that this is invalid C

File: tests/integration/test_tracking.py (hunk 6 above)
Comment (pablogsal) [reply tests/integration/test_tracking.py:740]: For some reason the high water mark was higher than expected and I didn't want to fix this in this PR

File: src/memray/_memray/native_resolver.cpp [hunk 7]
```diff
@@ -343,9 +343,10 @@ SymbolResolver::addSegments(
     // because the libbacktrace callback in findBacktraceState operates on char*
     const char* interned_filename = nullptr;
     auto filename_index = d_string_storage->internString(filename, &interned_filename);
+
     auto state = findBacktraceState(interned_filename, addr);
     if (state == nullptr) {
-        LOG(ERROR) << "Failed to prepare a backtrace state for " << filename;
+        LOG(DEBUG) << "Failed to prepare a backtrace state for " << filename;
```
[the remark this answers is not in the document]
Comment (pablogsal) [reply src/memray/_memray/native_resolver.cpp:349]: > Or was the level too high on Linux as well?

No, on linux is likely an error. I will fix this

File: tests/integration/test_tracking.py (hunk 6 above)
Comment (pablogsal) [reply tests/integration/test_tracking.py]: Oh, turns out it doesn't fail anymore 🤷

I will remove the skip then :)

**Review by godlygeek — APPROVED** [review]
LGTM
